### PR TITLE
Support more ScalarFunction in Ballista

### LINKED ISSUE #2007: [Ballista] Support more ScalarFunction in Ballista
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
Now 
```
select * from lineitem where l_quantity = random()*10;
[2022-03-14T06:19:49Z INFO  ballista_core::execution_plans::distributed_query] Connecting to Ballista scheduler at http://localhost:50050
Internal("failed to serialize logical plan: General(\"Unsupported scalar function Random\")")
```
**Describe the solution you'd like**
Add more ScalarFunction in `datafusion.proto`, and test it.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by Ted-Jiang**
# Which issue does this PR close?

Closes #2007 .

 # Rationale for this change
Now support more `ScalarFunction` in ballista. like "random"
start 1 scheduler + 1 executor
run client with `cargo run --features=ballista   --bin datafusion-cli  -- --host localhost --port 50050`
Before:
```
select * from lineitem where l_quantity = random()*10;

Internal("failed to serialize logical plan: General(\"Unsupported scalar function Random\")")
```
Now:
```
select * from lineitem where l_quantity = trunc(random() * 9 + 1) limit 3;
+------------+-----------+-----------+--------------+------------+-----------------+------------+-------+--------------+--------------+------------+--------------+---------------+-------------------+------------+-------------------------+
| l_orderkey | l_partkey | l_suppkey | l_linenumber | l_quantity | l_extendedprice | l_discount | l_tax | l_returnflag | l_linestatus | l_shipdate | l_commitdate | l_receiptdate | l_shipinstruct    | l_shipmode | l_comment               |
+------------+-----------+-----------+--------------+------------+-----------------+------------+-------+--------------+--------------+------------+--------------+---------------+-------------------+------------+-------------------------+
| 7          | 145243    | 7758      | 2            | 9          | 11594.16        | 0.08       | 0.08  | N            | O            | 1996-02-01 | 1996-03-02   | 1996-02-19    | TAKE BACK RETURN  | SHIP       | es. instructions        |
| 35         | 120896    | 8433      | 3            | 7          | 13418.23        | 0.06       | 0.04  | N            | O            | 1996-01-19 | 1995-12-22   | 1996-01-29    | NONE              | MAIL       |  the carefully regular  |
| 163        | 192642    | 5162      | 4            | 5          | 8673.2          | 0.02       | 0     | N            | O            | 1997-11-17 | 1997-10-09   | 1997-12-05    | DELIVER IN PERSON | TRUCK      |  must belie             |
+------------+-----------+-----------+--------------+------------+-----------------+------------+-------+--------------+--------------+------------+--------------+---------------+-------------------+------------+-------------------------+
3 rows in set. Query took 1.897 seconds.
```

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion-proto/src/from_proto.rs [hunk 1]
```diff
@@ -992,32 +1017,86 @@ impl TryFrom<&protobuf::LogicalExprNode> for Expr {
...
+                        (&args[2]).try_into()?,
+                    )),
+                    ScalarFunction::Reverse => Ok(reverse((&args[0]).try_into()?)),
+                    ScalarFunction::Right => {
+                        Ok(right((&args[0]).try_into()?, (&args[1]).try_into()?))
+                    }
+                    // ScalarFunction::Rpad => Ok(rpad(vec![])),
+                    // ScalarFunction::RegexpReplace => {
+                    //     Ok(regexp_replace(vec![]))
+                    // }
```
Comment (Ted-Jiang) [inline datafusion-proto/src/from_proto.rs:1071]: Add  issue #2009

File: datafusion-proto/src/from_proto.rs [hunk 2]
```diff
@@ -992,32 +1017,86 @@ impl TryFrom<&protobuf::LogicalExprNode> for Expr {
...
+                        Ok(starts_with((&args[0]).try_into()?, (&args[1]).try_into()?))
+                    }
+                    ScalarFunction::Strpos => {
+                        Ok(strpos((&args[0]).try_into()?, (&args[1]).try_into()?))
+                    }
+                    ScalarFunction::Substr => {
+                        Ok(substr((&args[0]).try_into()?, (&args[1]).try_into()?))
+                    }
+                    ScalarFunction::ToHex => Ok(to_hex((&args[0]).try_into()?)),
+                    //ScalarFunction::ToTimestampMicros Ok(totime((&args[0]).try_into()?)),
```
Comment (Ted-Jiang) [inline datafusion-proto/src/from_proto.rs:1092]: Add issue #2010

File: datafusion-proto/src/from_proto.rs (hunk 1 above)
Comment (liukun4515) [reply datafusion-proto/src/from_proto.rs:1071]: Can we fix the #2009 and #2010  in this pull request?

Comment (yjshen) [reply datafusion-proto/src/from_proto.rs:1071]: I agree that we should implement these functions in this pr as much as possible. Only leave the controversial ones.

Comment (Ted-Jiang) [reply datafusion-proto/src/from_proto.rs:1071]: Sure, will find a path how to fix these.

Comment (yjshen) [reply datafusion-proto/src/from_proto.rs]: Please let me know if it's ready for the next round of review.

Comment (Ted-Jiang) [reply datafusion-proto/src/from_proto.rs]: @yjshen PTAL

File: datafusion-proto/src/from_proto.rs [hunk 3]
```diff
@@ -979,8 +1008,6 @@ impl TryFrom<&protobuf::LogicalExprNode> for Expr {
                     ScalarFunction::Sin => Ok(sin((&args[0]).try_into()?)),
                     ScalarFunction::Cos => Ok(cos((&args[0]).try_into()?)),
                     ScalarFunction::Tan => Ok(tan((&args[0]).try_into()?)),
-                    // ScalarFunction::Asin => Ok(asin(&args[0]).try_into()?)),
```
Comment (yjshen) [inline datafusion-proto/src/from_proto.rs:982]: I think the commented-out Asin, Acos, Concat, Array, Nullif, and md5 are removed?

Comment (Ted-Jiang) [reply datafusion-proto/src/from_proto.rs:982]: Nullif is not in `expr.rs`. will address in another PR.

Comment (Ted-Jiang) [reply datafusion-proto/src/from_proto.rs:982]: Opps,  forgot the deleted line, have added.

**Review by yjshen — APPROVED** [review]
Thanks @Ted-Jiang, LGTM.

**Pull request merged**